src/osp/id_registry.h

```diff
@@ -0,0 +1,341 @@
+
+    /**
+     * @brief Create multiple IDs
+     *
+     * @param out   [out] Iterator out
+     * @param count [in] Number of IDs to generate
+     */
+    template<typename IT_T>
+    void create(IT_T out, size_t count);
+
+    /**
+     * @return Size required to fit all existing IDs, or allocated size if
+     *         reserved ahead of time
+     */
+    constexpr size_t capacity() const noexcept { return m_deleted.size(); }
+
+    /**
+     * @return Number of registered IDs
+     */
+    constexpr size_t size() const { return capacity() - m_deleted.count(); }
+
+    /**
+     * @brief Allocate space for at least n IDs
+     *
+     * @param n [in] Number of IDs to allocate for
+     */
+    void reserve(size_t n)
+    {
+        m_deleted.resize(n, true);
+    }
+
+    /**
+     * @brief Remove an ID. This will mark it for reuse
+     *
+     * @param id [in] ID to remove
+     */
+    void remove(ID_T id)
+    {
+        if ( ! exists(id))
+        {
+            throw std::runtime_error("ID does not exist");
+        }
+
+        m_deleted.set(id_int_t(id));
+    }
+
+    /**
+     * @brief Check if an ID exists
+     *
+     * @param id [in] ID to check
+     *
+     * @return true if ID exists
+     */
+    bool exists(ID_T id) const noexcept
+    {
+        return ! m_deleted.test(id_int_t(id));
+    };
+
+    template <typename FUNC_T>
+    void for_each(FUNC_T func);
+
+private:
+    HierarchicalBitset<uint64_t> m_deleted;
+
+}; // class IdRegistry
+
+template<typename ID_T, bool NO_AUTO_RESIZE>
+template<typename IT_T>
+void IdRegistry<ID_T, NO_AUTO_RESIZE>::create(IT_T out, size_t count)
+{
+    if (m_deleted.count() < count)
+    {
+        // auto-resize
+        if constexpr (NO_AUTO_RESIZE)
+        {
+            throw std::runtime_error(
+                    "Reached max capacity with automatic resizing disabled");
+        }
+        else
+        {
+            reserve(std::max(capacity() + count, capacity() * 2));
+        }
+    }
+
+    m_deleted.take<IT_T, ID_T>(out, count);
+}
+
+template<typename ID_T, bool NO_AUTO_RESIZE>
+template <typename FUNC_T>
+void IdRegistry<ID_T, NO_AUTO_RESIZE>::for_each(FUNC_T func)
+{
+    // TODO: this is kind of inefficient and doesn't utilize the hierarchy
+    for (size_t i = 0; i < size(); i ++)
+    {
+        if ( ! m_deleted.test(i) )
+        {
+            func(ID_T(i));
+        }
+    }
+}
+
+//-----------------------------------------------------------------------------
+
+/**
+ * @brief Long term reference counted storage for IDs
+ */
+template<typename ID_T, typename REG_T>
+class IdStorage
+{
+    friend REG_T;
+
+public:
+    IdStorage() : m_id{ id_null<ID_T>() } { }
+    ~IdStorage() { assert( ! has_value() ); }
+
+    // Delete copy
+    IdStorage(IdStorage const& copy) = delete;
+    IdStorage& operator=(IdStorage const& copy) = delete;
+
+    // Allow move only if there's no ID stored
+    IdStorage(IdStorage&& move)
+    {
+        assert( ! has_value() );
```

Comment:
is this a c assert?

i feel like we probably need to grow a more elaborated assert mechanism.

E.g. at my work we have an assert framework that can do a bunch of things like

1. Pop up a dialog box if in a debug build, and depending on which button is pressed, launch the system debugger to that specific line of code (There's an intel CPU instruction that just does this by magic, somehow).
2. Optionally throw an exception on assert failure
3. Optionally terminate the program

Reply:
Yeah that's a plain C assert. I read that it was quite dangerous to throw exceptions in the destructor. I'd often see a macro used for asserts (how about OSP_ASSERT(...)?), though I'm not sure what they're defined as.